### ISSUE TITLE
String value is being mangled

### ISSUE THREAD

**Issue by rad-pat** [qooxdoo/qooxdoo-compiler#715]
In my class, I have a static defined such as:
```
statics: {
      ROOT_NODE: "__root__"
}
```
When it is compiled, the value is mangled to e.g. `ROOT_NODE: "__root____P_131_0"` . The value is a constant which represents a value in the backend so this should not be mangled by the compile process.

**Comment by hkollmann** [comment]
@johnspackman:
any idea why string literals are mangled?

ClassFile.js#1284: 
          Literal(path) {
            if (typeof path.node.value == "string") {
              path.node.value = t.encodePrivate(path.node.value);
            }
          },

**Comment by johnspackman** [comment]
For private variable support; eg there is code in the framework considered legitimate like this.dispose("__myPrivateVar")

The generator uses some better heuristics to detect what should not be mangled I guess; either that or a multiple pass scan, or maybe it special cases 'dispose'

**Comment by hkollmann** [comment]
Here the part from the generator:

def lookup(id, node, privates, globalPrivs):
    # privates = { "<private>" : "<repl>", ... }
    name = None
    
    if node.type == "definition":
        if node.getChild("identifier", False):
            name = node.getChild("identifier").get("value", '')
        # treat node.getChild("assignment", False) later on its own

    elif node.type == "keyvalue":
        name = node.get("key", '')
        
    elif node.type == "assignment":
        lval = node.children[0]
        if lval.isVar():
            if lval.type == "identifier":
                name = lval.get("value")
            elif lval.type == "dotaccessor":
                last = lval.getRightmostOperand()
                name = last.get("value")
        
    if name and name.startswith("__") and not name in privates:
        privates[name] = crypt(id, name, globalPrivs)
        
        #if not name in used:
        #    used[name] = [id]
        #elif not id in used[name]:
        #    used[name].append(id)

    if node.hasChildren():
        for child in node.children:
            lookup(id, child, privates, globalPrivs)
        
    return privates

**Comment by johnspackman** [comment]
that code seems to just be saying that anything beginning with `__` is mangled (unless it is already mangled); but there must be something else which prevents `__root__` from being mangled.

The issue is how do we know that the symbol "__xxx" is intended to be private, as opposed to just a symbol which has two underscores at the front?  The compiler takes the point of view that all symbols are supposed to be known (ie that to have undeclared private variables is poor practice) which means that we can reliably conclude that `__xxx` can be mapped to `__whateverWeLike__xxx` reliably.

The issue comes with strings - because code can use string lookups to refer to private members (eg calls to `this._disposeObjects("__xxx")` in destructors) it is necessary to map them also.  

I've picked a few source code files at random that have privates in strings, and it seems that these references happen only after the variable is declared; my guess is that the generator relies on variable declarations to determine whether to mangle strings, rather than pattern matching.  If my guess is correct, this would mean that the generator will mangle (or not) depending on the order that the code is written, leading to inconsistency for example:

```
  construct: function() {
    this.base(arguments);
    this.doSomething("__myPrivate"); // not mangled
    this.doMore();
  },

  members: {
    doMore: function() {
      this.doSomething("__myPrivate"); // Is mangled
    }
  }
```

IMHO it is unreasonable and unexpected to be mangling string values, but this is something which we have to support for legacy reasons - I'd like to deprecate that at some point in the future, but that's probably another issue altogether (not least because it would affect every `_disposeObjects` call, although OTOH many of those calls are irrelevant now that there is automatic memory management)

So the fix would be to see a string starting with "__", and if it is not a known private identifier then prevent it from being mangled anywhere in that class file.  Detecting known identifiers would be done by collecting `Property` nodes and looking at their `key` where it is of type `"Identifier"` - possibly restricting this to only check the immediate descendants of `statics` and `members` in the class declaration

**Comment by johnspackman** [comment]
fixed in https://github.com/qooxdoo/qooxdoo-compiler/pull/785
